File: backend/graph/tools/NonFactors/FactorSecTradeAgg.py
import numpy as np
import math
from L3FactorFrame.FactorBase import FactorBase
from L3FactorFrame.tools.DecimalUtil import isEqual, notEqual
from L3FactorFrame.tools.helper_functions import trade_field_agg


class FactorSecTradeAgg(FactorBase):
    def __init__(self, config, factorManager, marketDataManager):
        super().__init__(config, factorManager, marketDataManager)
        self.trade_buy_money_list = []
        self.trade_sell_money_list = []
        self.trade_buy_num_list = []
        self.trade_sell_num_list = []
        self.trade_buy_volume_list = []
        self.trade_sell_volume_list = []


    def calculate(self):
        timestamps = self.getPrevTick("Timestamp")
        sample_1s_flag = self.get_sample_1s_flag()
        if sample_1s_flag == 1:
            # sample_1s_flag表示超过这一秒的第一条数据，将end_timestamp设置为这一秒的整秒，作为右闭区间
            # timestamps向上取整为当前时间
            trade_qtys = self.getPrevSecTrade("Volume", 1, end_timestamp = math.ceil(timestamps))
            trade_moneys = self.getPrevSecTrade("Amount", 1, end_timestamp = math.ceil(timestamps))
            trade_sides = self.getPrevSecTrade("BSFlag", 1, end_timestamp = math.ceil(timestamps))

            active_buy_money, active_sell_money, active_buy_num, active_sell_num, active_buy_volume, active_sell_volume = \
                trade_field_agg(trade_qtys, trade_moneys, trade_sides)

            self.trade_buy_money_list.append(active_buy_money)
            self.trade_sell_money_list.append(active_sell_money)
            self.trade_buy_num_list.append(active_buy_num)
            self.trade_sell_num_list.append(active_sell_num)
            self.trade_buy_volume_list.append(active_buy_volume)
            self.trade_sell_volume_list.append(active_sell_volume)

        # Nonfactor的结果不用存
        self.addFactorValue(0.0)

File: backend/graph/domain/codegen/tools/cpp_nonfactor/FactorSecTradeAgg.cpp
#include "huatai/atsquant/factor/SAMPLE_1S/nonfactor/FactorSecTradeAgg.h"

namespace huatai::atsquant::factor {

  void FactorSecTradeAgg::on_init() { set_sample_mode(false); }


Status FactorSecTradeAgg::calculate() {
  auto seq_no = get_market_data().get_prev_n_quote(1).item<QuoteSchema::Index::TRIGGER_APPL_SEQ_NUM>();
  auto sample_1s_flag = get_sample_flag();

  if (sample_1s_flag && get_factor_manager_param().calculate_mode == FactorManagerParam::CalculateMode::SAMPLE_1S) {
    auto last_quote = get_market_data().get_prev_n_quote(1)[0];
    // 获取采样的时间戳，为上一秒的整秒+1秒
    auto sec_mdtime = (last_quote.item<QuoteSchema::Index::TRIGGER_TIME>() - 1) / 1000 * 1000 + 1000L;
    if (trigger_time_list.size() != 0 and sec_mdtime <= trigger_time_list.back()) {
      sec_mdtime = trigger_time_list.back() + 1000L;
    }
    trigger_time_list.push(sec_mdtime); // 这一整秒的最后一条

    auto trades = get_market_data().get_prev_n_ms_trade(1000, sec_mdtime);
    // auto trades = get_market_data().get_prev_n_ms_trade(1000);
    auto trade_len = trades.len();
    if (trade_len > 0) {
      // 区间范围为上一秒的所有成交数据
      const auto *trade_qtys = trades.item_address<TradeSchema::Index::QUANTITY>();
      const auto *trade_moneys = trades.item_address<TradeSchema::Index::TURNOVER>();
      const auto *trade_sides = trades.item_address<TradeSchema::Index::SIDE>();

      auto active_buy_money = 0.0;
      auto active_sell_money = 0.0;
      auto active_buy_num = 0.0;
      auto active_sell_num = 0.0;
      auto active_buy_volume = 0.0;
      auto active_sell_volume = 0.0;
      auto active_buy_price = 0.0;
      auto active_sell_price = 0.0;

      trades.for_each([&](const RowView<TradeSchema> &rv) {
        trade_price_list.push(rv.item<TradeSchema::Index::PRICE>());
        trade_side_list.push(rv.item<TradeSchema::Index::SIDE>());
        trade_quantity_list.push(rv.item<TradeSchema::Index::QUANTITY>());
        if (rv.item<TradeSchema::Index::SIDE>() == 1) {
          // active_buy_money += rv.item<TradeSchema::Index::TURNOVER>();
          active_buy_money += rv.item<TradeSchema::Index::QUANTITY>() * rv.item<TradeSchema::Index::PRICE>();
          active_buy_volume += rv.item<TradeSchema::Index::QUANTITY>();
          active_buy_num++;
          active_buy_price += rv.item<TradeSchema::Index::PRICE>();
        } else if (rv.item<TradeSchema::Index::SIDE>() == 2) {
          // active_sell_money += rv.item<TradeSchema::Index::TURNOVER>();
          active_sell_money += rv.item<TradeSchema::Index::QUANTITY>() * rv.item<TradeSchema::Index::PRICE>();
          active_sell_volume += rv.item<TradeSchema::Index::QUANTITY>();
          active_sell_num++;
          active_sell_price += rv.item<TradeSchema::Index::PRICE>();
        }
      });

      trade_buy_money_list.push(active_buy_money);
      trade_sell_money_list.push(active_sell_money);
      trade_buy_num_list.push(active_buy_num);
      trade_sell_num_list.push(active_sell_num);
      trade_buy_volume_list.push(active_buy_volume);
      trade_sell_volume_list.push(active_sell_volume);
      trade_net_volume_list.push(active_buy_volume - active_sell_volume);
      trade_buy_price_list.push(active_buy_price);
      trade_sell_price_list.push(active_sell_price);
    } else {
      trade_buy_money_list.push(0.0);
      trade_sell_money_list.push(0.0);
      trade_buy_num_list.push(0.0);
      trade_sell_num_list.push(0.0);
      trade_buy_volume_list.push(0.0);
      trade_sell_volume_list.push(0.0);
      trade_net_volume_list.push(0.0);
      trade_buy_price_list.push(0.0);
      trade_sell_price_list.push(0.0);
    }
  }
  return Status::OK();
}

} // namespace huatai::atsquant::factor
